public/app/features/alerting/unified/components/rules/state-history/LogRecordViewer.tsx

```diff
@@ -68,30 +91,84 @@ export const LogRecordViewerByTimestamp = memo(
               id={key.toString(10)}
               key={key}
               data-testid={key}
-              ref={(element) => element && timestampRefs.set(key, element)}
+              ref={(element) => {
+                if (element) {
+                  timestampRefs.current.set(key, element);
+                } else {
+                  timestampRefs.current.delete(key);
+                }
+              }}
               className={styles.listItemWrapper}
             >
               <Timestamp time={key} />
-              <div className={styles.logsContainer}>
-                {records.map(({ line }) => (
-                  <Fragment key={uniqueId()}>
-                    <AlertStateTag state={line.previous} size="sm" muted />
-                    <Icon name="arrow-right" size="sm" />
-                    <AlertStateTag state={line.current} />
-                    <Stack>{line.values && <AlertInstanceValues record={line.values} />}</Stack>
-                    <div>
-                      {line.labels && (
-                        <TagList
-                          tags={omitLabels(Object.entries(line.labels), commonLabels).map(
-                            ([key, value]) => `${key}=${value}`
-                          )}
-                          onClick={onLabelClick}
-                        />
-                      )}
+              {records.map(({ line }, idx) => {
+                const id = line.fingerprint ?? `${key}-${idx}`;
+                const errorText = line.error ?? '';
+                const isMultiline = errorText.includes('\n');
+                const isTruncated = Boolean(isTruncatedById[id]);
+                const isExpanded = Boolean(expandedErrorIds[id]);
+                const shouldShowToggle = isMultiline || isTruncated || isExpanded;
+                const isErrorRow =
+                  mapStateWithReasonToBaseState(line.current) === GrafanaAlertState.Error && Boolean(line.error);
+                return (
+                  <Fragment key={id}>
+                    <div className={styles.logsContainer}>
+                      <AlertStateTag state={line.previous} size="sm" muted />
+                      <Icon name="arrow-right" size="sm" />
+                      <AlertStateTag state={line.current} />
+                      <Stack>{line.values && <AlertInstanceValues record={line.values} />}</Stack>
+                      <div>
+                        {line.labels && (
+                          <TagList
+                            tags={omitLabels(Object.entries(line.labels), commonLabels).map(
+                              ([key, value]) => `${key}=${value}`
+                            )}
+                            onClick={onLabelClick}
+                          />
+                        )}
+                      </div>
                     </div>
+                    {isErrorRow && (
```

Comment:
Let's turn this into a separate component so we can re-use that in the Central Alert State history view :D